poetry/core/masonry/builders/builder.py

```diff
@@ -113,23 +129,53 @@ def is_excluded(self, filepath):  # type: (Union[str, Path]) -> bool
 
         return False
 
-    def find_files_to_add(self, exclude_build=True):  # type: (bool) -> list
+    def find_files_to_add(
+        self, exclude_build=True
+    ):  # type: (bool) -> Set[BuildIncludeFile]
         """
         Finds all files to add to the tarball
         """
-        to_add = []
+        to_add = set()
 
         for include in self._module.includes:
+            include.refresh()
+            formats = include.formats or ["sdist"]
+
             for file in include.elements:
                 if "__pycache__" in str(file):
                     continue
 
                 if file.is_dir():
+                    if self.format in formats:
+                        for current_file in file.glob("**/*"):
+                            include_file = BuildIncludeFile(
+                                path=current_file, source_root=self._path
+                            )
+
+                            if (
+                                include_file.relative_to_source_root() not in to_add
```

Comment:
Shouldn't this be
```suggestion
                                include_file not in to_add
```

Also, since this is a set we can skip this check.